mesonbuild/build.py

```diff
@@ -2091,6 +2092,15 @@ def __init__(
             compilers: T.Dict[str, 'Compiler'],
             kwargs):
         self.prelink = T.cast('bool', kwargs.get('prelink', False))
+        if StaticLibrary.allow_thin_archives and 'cuda' in environment.coredata.compilers[for_machine]:
+            # nvcc chokes on thin archives:
+            #   nvlink fatal   : Could not open input file 'libfoo.a.p'
+            #   nvlink fatal   : elfLink internal error
+            # hence we disable them if 'cuda' is enabled globally. See also
+            # - https://github.com/mesonbuild/meson/pull/9453
+            # - https://github.com/mesonbuild/meson/issues/9479#issuecomment-953485040
+            mlog.debug('cuda enabled globally, disabling thin archives, since nvcc/nvlink cannot handle thin archives natively')
+            StaticLibrary.allow_thin_archives = False
```

Comment:
I don't think this is the right place to put this information, I think this belongs in the backend.

One reason for that, is that this can be changed by a subproject or by `add_language()` call, so you could initially generate targets with the assumption that thin archives are allowed, then end up disabling them. I know you're not doing that now, but I'm concerned that we'll end up with hard to solve bugs later.

I also think this should be done on a per-machine basis, no reason to turn off thin archives on the build machine if cuda is only added for the host machine.

So, it makes more sense to me to do this check in the Backend class, store the information in the backend, and use it there.